nullaway/src/main/java/com/uber/nullaway/AnnotatedTypeWrapper.java

```diff
@@ -0,0 +1,11 @@
+package com.uber.nullaway;
+
+import com.sun.source.tree.Tree;
+import java.util.HashSet;
+
+public interface AnnotatedTypeWrapper<T1, T2> {
```

Comment:
So I think the overall code structure should look like this roughly:
```java
void checkIdenticalInstantiations(Tree lhsTree, Tree rhsTree, ...) {
  TypeWrapper lhsWrapper = getTypeWrapper(lhsTree);
  TypeWrapper rhsWrapper = getTypeWrapper(rhsTree);
  checkIdenticalWrappers(lhsWrapper, rhsWrapper);
}

void checkIdenticalWrappers(TypeWrapper lhsWrapper, TypeWrapper rhsWrapper) {
  if (lhsWrapper == null || rhsWrapper == null) return;
  if (!ASTHelpers.isSameType(lhsWrapper.getWrapped(), rhsWrapper.getWrapped()) {
    rhsWrapper = rhsWrapper.viewAsSupertype(lhsWrapper.getWrapped());
  }
  if (!sameNullableTypeIndices(lhsWrapper,rhsWrapper)) reportError(...);
  List<TypeWrapper> lhsNested = lhsWrapper.getWrappersForNested();
  List<TypeWrapper> rhsNested = rhsWrapper.getWrappersForNested();
  for (int i = 0; i < lhsNested.size(); i++) {
    checkIdenticalWrappers(lhsNested.get(i), rhsNested.get(i));
  }
}

@Nullable TypeWrapper getTypeWrapper(Tree tree) {
  // logic based on the type of the tree, and return null if
  // it's not for a generic type or doesn't need to be checked
}

interface TypeWrapper {
  // the underlying type being wrapped
  Type getWrapped();
  List<Integer> getNullableTypeIndices();
  List<TypeWrapper> getWrappersForNested(); // can have null elements
}
```

The key differences from your current code structure are:

* The `AnnotatedTypeWrapper` functions should need to take in the thing to be wrapped as an argument.  It should just be passed to the constructor and stored in a field.
* The logic to check for an assignment type match should be written **once** outside the type wrappers, not twice, once inside each type wrapper.  This way we avoid duplication of logic.  That checking logic should rely on the `AnnotatedTypeWrapper` APIs for all logic related to the types, which is why in the example above I added a method `getWrappersForNested()` to get the nested type wrappers.

Does this make sense?  If so please refactor the code to match this structure.